raster-foundry/raster-foundry
# Add task summary field to annotation projects table; Add DB trigger for operations on tasks table

## Overview

This PR:
- adds a new field `task_status_summary` to the `annotation_projects` table to store task status distribution for annotation projects
- adds a database trigger on `INSERT`/`UPDATE`/`DELETE` operations on `tasks` table so that `task_status_summary` in `annotation_projects` table is in sync
- adds back the old `authQuery` logic so that the `counts` field should work the same as before
- fixes a bug in the calculation of `hasNext` field in `Dao`

### Checklist

- ~Swagger specification updated~ (There is no spec for annotation project set of endpoints)
- [X] New tables and queries have appropriate indices added (a `gin` index is added to `task_status_summary` field, in case we are adding jsonb field sorting)
- ~Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~
- [X] Any new SQL strings have tests (modified existing tests)
- ~Any new endpoints have scope validation and are included in the integration test csv~

## Testing Instructions

- Tests are updated to test the new trigger and new column, so CI should pass
- Run migrations
- `api/assembly` and spin up servers
- If your local DB has and only has these two annotation projects: `Florence - Alabama` and `Sample Project`, then the below steps should pass for you; if your local DB annotation projects are different, run similar tests as below:
     - `GET` to `api/annotation-projects` should list 2 annotation projects
     - `GET` to `api/annotation-projects?taskStatusesInclude=VALIDATED` should list only `Florence - Alabama` project
     - `GET` to `api/annotation-projects?taskStatusesInclude=UNLABELED` should list only `Sample Project` project
     - make sure the `hasNext` field of each of the above return reflect the reality
     - make sure the `count` field of each of the above return reflect the reality
- Choose a task from a project. Check the summary field count of a project by `GET` to `api/annotation-projects/<id>`. Update its status by sending the updated task object to `api/annotation-projects/<id>/tasks/<id>`.  Check the project's status again.
- Perform similar operation as the above, but delete a task this time. Check project task status before and after the deletion.

Closes https://github.com/raster-foundry/raster-foundry/issues/5377
